Eventtrack accessibility update

reichlfl opened the pull request:
This PR enables basic accessibility information for the thread bars (EventTrack, TracepointTrack, ThreadStateTrack), and updates E2E tests to verify the newly available information:

- Tracks now report the existence and locations of thread bars underneath, but not individual samples
- Tracks expose the existence of timers: an additional TimerPane is reported if at lease one timer is visible in the current viewport, but no individual timers are reported
- The E2E tests for thread states, instrument function and frame track check for the respective content in the capture window

Suggest to go through it commit by commit as the first few commits lay some ground work.

dpallotti commented on contrib/automation_tests/test_cases/capture_window.py:
```diff
@@ -239,3 +242,42 @@ def _execute(self, length_in_seconds: int = 5, collect_thread_states: bool = Fal
+class CheckThreadStates(CaptureWindowE2ETestCaseBase):
+    def _execute(self, track_name_contains: str, expect_exists: bool = True):
+        tracks = self._find_tracks(track_name_contains)
+        self.expect_true(len(tracks) > 0, 'Found tracks matching %s' % track_name_contains)
```
I'd wrap `%s` in quotes. Here and in the classes below.

dpallotti commented on contrib/automation_tests/test_cases/capture_window.py:
```diff
@@ -239,3 +242,42 @@ def _execute(self, length_in_seconds: int = 5, collect_thread_states: bool = Fal
+    def _execute(self, track_name_contains: str, expect_exists: bool = True):
+        tracks = self._find_tracks(track_name_contains)
+        self.expect_true(len(tracks) > 0, 'Found tracks matching %s' % track_name_contains)
+        logging.info('Checking for threadstates in %s tracks', len(tracks))
```
```suggestion
        logging.info('Checking for thread states in %s tracks', len(tracks))
```

dpallotti commented on contrib/automation_tests/test_cases/capture_window.py:
```diff
@@ -239,3 +242,42 @@ def _execute(self, length_in_seconds: int = 5, collect_thread_states: bool = Fal
+        for track in tracks:
+            track = Track(track)
+            if expect_exists:
+                self.expect_true(track.thread_states is not None, 'Track %s has a threadstate pane' % track.name)
```
```suggestion
                self.expect_true(track.thread_states is not None, 'Track %s has a thread state pane' % track.name)
```

dpallotti commented on contrib/automation_tests/test_cases/capture_window.py:
```diff
@@ -239,3 +242,42 @@ def _execute(self, length_in_seconds: int = 5, collect_thread_states: bool = Fal
+            if expect_exists:
+                self.expect_true(track.thread_states is not None, 'Track %s has a threadstate pane' % track.name)
+            else:
+                self.expect_true(track.thread_states is None, 'Track %s has a no threadstate pane' % track.name)
```
```suggestion
                self.expect_true(track.thread_states is None, 'Track %s has a no thread state pane' % track.name)
```

dpallotti commented on src/OrbitGl/EventTrack.cpp:
```diff
@@ -32,8 +32,10 @@ using orbit_client_protos::CallstackEvent;
-    : ThreadBar(app, time_graph, layout, capture_data, thread_id), color_{0, 255, 0, 255} {}
+                       const CaptureData* capture_data, ThreadID thread_id,
+                       CaptureViewElement* parent)
+    : ThreadBar(app, time_graph, layout, capture_data, thread_id, parent, "Events"),
```
Seeing `"Events"` passed as `name` and used in the E2E tests reminded me that `CallstackTrack` and `"Callstacks"` would be more appropriate. Just throwing it there...

reichlfl replied:
I was going to rename those classes to not be called "Track" anymore, I think `CallstackThreadBar` would be more fitting given the base class is called `ThreadBar`. But I wanted to do this in a follow-up PR to make the review easier.

dpallotti reviewed (approved):
Minor comments. Thanks for the change!

reichlfl commented on src/OrbitGl/CaptureViewElement.h:
```diff
@@ -35,6 +35,11 @@ class CaptureViewElement : public Pickable {
   [[nodiscard]] Vec2 GetSize() const { return size_; }
 
+  [[nodiscard]] orbit_accessibility::AccessibleInterface* GetOrCreateAccessibleInterface();
+  [[nodiscard]] const orbit_accessibility::AccessibleInterface* GetAccessibleInterface() const {
+    return accessible_interface_.get();
+  }
```
Exactly, it's only for the `const` case. Accessible interfaces may need this to retrieve the parent-interface in `AccessibleParent`, which is const and can safely assume that the parent has already created the interface since the tree is always traversed from the root. And in this case, it also needs to be public.

reichlfl commented on src/OrbitGl/AccessibleThreadBar.cpp:
```diff
@@ -0,0 +1,56 @@
+  Vec2 pos = thread_bar_->GetPos();
+
+  // TODO(b/177350599): This could be cleaned up with clearer coordinate systems
+  if (thread_bar_->GetParent() != nullptr) {
```
Yes, kind of... I think the core issue is that we're specifying absolute positions when it would make more sense to specify positions relative to the parent, then `pos` would be correct in both cases.

This can't actually happen since we're parenting all elements under `TimeGraph` currently, I just wanted to make sure this case works and is well-defined to have it unit-testable (when tracks are finally testable).

florian-kuebler commented on src/OrbitGl/AccessibleTimeGraph.cpp:
```diff
@@ -30,7 +30,9 @@ int TimeGraphAccessibility::AccessibleChildCount() const {
 
 const AccessibleInterface* TimeGraphAccessibility::AccessibleChild(int index) const {
-  return time_graph_->GetTrackManager()->GetVisibleTracks()[index]->AccessibilityInterface();
+  return time_graph_->GetTrackManager()
```
Why is this reformatted? Was that intended?

reichlfl replied:
Yes, the function name changed from `AccessibilityInterface` to `GetOrCreateAccessibilityInterface`

florian-kuebler reviewed (approved):
Thanks for that change, in general looks pretty good. I guess the most relevant points are pointed out by @dpallotti.

reichlfl commented on src/OrbitGl/AccessibleThreadBar.cpp:
```diff
@@ -0,0 +1,56 @@
+  if (thread_bar_->GetParent() != nullptr) {
+    Vec2 parent_pos = thread_bar_->GetParent()->GetPos();
+    pos[0] -= parent_pos[0];
+    pos[1] = parent_pos[1] - pos[1];
```
Right, that's better in terms of readability - changed accordingly, thanks!

reichlfl commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -17,25 +17,35 @@
+int AccessibleTrackContent::AccessibleChildCount() const {
+  std::vector<CaptureViewElement*> children = track_->GetVisibleChildren();
+  int result = static_cast<int>(children.size());
+  return result;
```
This was changed in a later commit anyways... I cleaned it up nevertheless.

reichlfl commented on src/OrbitGl/TimerTrack.cpp:
```diff
@@ -98,6 +98,8 @@ void TimerTrack::DrawTimer(const TextBox* prev_text_box, const TextBox* next_tex
   if (current_timer_info.start() >= *min_ignore && current_timer_info.end() <= *max_ignore) return;
   if (!TimerFilter(current_timer_info)) return;
 
+  ++visible_timer_count_;
```
True, that's definitely better. Completely agree, changed.

reichlfl commented:
Thanks for the comments, @dpallotti and @florian-kuebler. I addressed all suggested fixes except the renaming, will do in another PR. I rebased and reordered commits, if you want to have another look it should be sufficient to take a look at the fixup! commits. In any case, I'd merge later today since you already approved and the fixes were straightforward, unless there's any complaints.
